### PR TITLE
www-apps/rutorrent: bump version to 4.0.2 and update 9999

### PR DESCRIPTION
**Pull request by henning-schild**
Note that rutorrent-9999.ebuild is now just a copy of the latest version.

Closes: https://bugs.gentoo.org/730206
Signed-off-by: Henning Schild <henning@hennsch.de>

### CODE REVIEW

**Comment by CyberGuerro** [comment]
I've tried a clean install that it seem ok, but when I open it on browser I get these errors:
```
[15.01.2023 10:13:28] WebUI started.
[15.01.2023 10:13:28] Bad response from server: (200 [parsererror,getplugins]) TypeError: theWebUI.createSelectableLabelElement is not a function
[15.01.2023 10:13:29] _cloudflare plugin: cloudscraper module can't be loaded in Python
[15.01.2023 10:14:22] XMLRPC Error: Method 'dht_statistics' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_check_hash' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_bind' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_dht_port' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_directory' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_download_rate' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_hash_interval' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_hash_max_tries' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_hash_read_ahead' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_http_cacert' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_http_capath' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_http_proxy' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_ip' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_max_downloads_div' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_max_downloads_global' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_max_file_size' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_max_memory_usage' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_max_open_files' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_max_open_http' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_max_peers' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_max_peers_seed' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_max_uploads' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_max_uploads_global' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_min_peers_seed' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_min_peers' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_peer_exchange' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_port_open' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_upload_rate' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_port_random' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_port_range' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_preload_min_size' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_preload_required_rate' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_preload_type' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_proxy_address' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_receive_buffer_size' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_safe_sync' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_scgi_dont_route' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_send_buffer_size' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_session' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_session_lock' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_session_on_completion' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_split_file_size' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_split_suffix' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_timeout_safe_sync' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_timeout_sync' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_tracker_numwant' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_use_udp_trackers' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_max_uploads_div' not defined [getsettings]
[15.01.2023 10:14:22] XMLRPC Error: Method 'get_max_open_sockets' not defined [getsettings]
```
I think that it needs something other packages or configuration.

Reverted to 3.10

**Comment by henning-schild** [comment]
@CyberGuerro did it work at all? Some plugins missing some rdeps would not surprise me, the ebuild does not mention them and you get warnings in the webui and some plugins might just not work.

And when you say "clean install" you will have to delete the "share"  (make a backup in case you care) folder and its subfolders of the "webapp". So you really get a clean start. And i also found that there is caching going on of some javascript code, so you should try a browser in "private mode" first and if that works ... clear browser caches so you do not get a combination of old caches stuff and new stuff.

**Comment by CyberGuerro** [comment]
I did all the steps, but nothing change...    ..I hope that some other one can test it. meanwhile I'll stay with my 3.10 package

**Comment by henning-schild** [comment]
> I did all the steps, but nothing change... ..I hope that some other one can test it. meanwhile I'll stay with my 3.10 package

Please try to give me an advice on how to repro the issue you see here so i can fix it. If you just say that it does not work for you you stall this pull request and 3.10 is not a valid escape since it is not in the tree and never has been. Here we talk about moving from 3.8 to 4.0 ... possibly skipping 3.10.

**Comment by henning-schild** [comment]
https://github.com/Novik/ruTorrent/issues/2302 this could be related, i found it by looking for "theWebUI.createSelectableLabelElement is not a function" ... unfortunately it does not give a clue on what might be wrong, just a strong suggestion on clear browser caches. So please try to clear those caches or use a "private tab".

If that works we can later try and follow up with upstream why the caches do keep carrying potentially old files even if the web-app was updated.

When writing that new ebuild i also ran into caching issues where "private tab" was the only escape, might just be an upstream issue to deal with later.

When looking at things like left around `.gitignore` it clearly seems like upstream tests their "master" better than their releases.

**Comment by CyberGuerro** [comment]
Good news:
Opened in anonymous tab and then in default tab no more errors.
I noticed that there are no USE flags, I suggest you to add this:
`cloudscraper? ( dev-python/cloudscraper )`
because there is a plugin "_cloudflare" that depend on it

and that's it.
I'll wait for your next release.
meanwhile I adding USE flag in my local ebuild.

**Comment by stickz** [comment]
Hi @henning-schild I'm one of the ruTorrent collaborators. It's required to clear your web browser cache, when upgrading from version 4 stable to version 3.10. Version 3 is unsupported and less polished. You want to be on version 4. 

All versions of ruTorrent have caching issues. There is no reason to stay on a particular version because of this issue. 

I disabled caching the web interface settings in the web browser for the next release. It's already cached on the web server, so there is no point in causing client-side caching issues. The task is asynchronous now, so the extra round trip doesn't matter.

**Comment by henning-schild** [comment]
> Good news: Opened in anonymous tab and then in default tab no more errors. 

Good to hear that, so it was some kind of caching issue in your browser. Maybe something the webserver or the gentoo webapp support could have resolved ... but let us deal with that on the next update.

> I noticed that there are no USE flags, I suggest you to add this: `cloudscraper? ( dev-python/cloudscraper )` because there is a plugin "_cloudflare" that depend on it and that's it. I'll wait for your next release. meanwhile I adding USE flag in my local ebuild.

That USE-flag was never in the gentoo tree so i kept it out for now, we can see about adding it later on. In fact i think an "optfeature" might be the better way to not invent new USE-flags. You will see that plugin not working with a sort of meaningful message, and you can install dev-python/cloudscraper on your own until rutorrent maybe helps you pull that in.

> PS: could you verify if **dev-php/maxmind-db-reader** is a dependency or no? I ask you this because when I installed rutorrent 3.10 I had to install that package directly. It should be good remove it from my world file due to explicit dependency of new rutorrent ebuild

The machine i have that new rutorrent on does not have that installed. I will assume it is not needed or just another potential USE-flag or optfeature candidate for some plugin which is "optional". Remove it and get back if that causes any core-functionality issues. If it is plugin-related, wait for the merge and we improve later.

**Comment by henning-schild** [comment]
> Hi @henning-schild I'm one of the ruTorrent collaborators. It's required to clear your web browser cache, when upgrading from version 4 stable to version 3.10. Version 3 is unsupported and less polished. You want to be on version 4.
> 
> All versions of ruTorrent have caching issues. There is no reason to stay on a particular version because of this issue.
> 
> I disabled caching the web interface settings in the web browser for the next release. It's already cached on the web server, so there is no point in causing client-side caching issues. The task is asynchronous now, so the extra round trip doesn't matter.
> 
> **Edit:** I also just changed the max age of the cache to 1 day for the next release to mitigate this problem. There is very little benefit in caching content for extended periods of time. It won't have any real world impact on web server load.
> 
> **Edit 2:** The recommended course of action is to wait until a **version 4.0.1** hot fix is released before pushing this to Gentoo Linux. This will give us time to thoroughly test a bunch of bug fixes and mitigation of caching issues. New features are frozen until **v4.1**.

partially understood ... the caching issues are known and being worked on and we should wait for an upcoming release

I will pause this effort and wait for you to maybe get back, or i will look at it every other week

Feel free to post a link to a PR or issue that we are waiting for, i assume you tell all your users and redistributors about that somewhere ...

**Comment by henning-schild** [comment]
@stickz can you maybe point me to some usage stats and other distros shipping rutorrent?

i hate to admit that i was very close to proposing https://github.com/jesec/flood and ditching https://github.com/Novik/ruTorrent/ but for rutorrent we already have something going on in gentoo, and we can still add flood later, as alternative or replacement

to me rutorrent seems not wide-spread, but as long as it is maintained and works i do not mind maintaining it in gentoo ... we still have to see if all of this ever gets merged though

**Comment by stickz** [comment]
> I will pause this effort and wait for you to maybe get back, or i will look at it every other week

I'll get back to you in two to four weeks, once upstream improvements are released. We want to incorporate as many fixes into version 4.0.1 as possible to improve stability. 

> to me rutorrent seems not wide-spread, but as long as it is maintained and works i do not mind maintaining it in gentoo ... we still have to see if all of this ever gets merged though

ruTorrent is maintained. The version 4.0 release is a complete overhaul of the ruTorrent web client, with ground breaking improvements. We spent over a year testing the changes before marking it as stable.

> @stickz can you maybe point me to some usage stats and other distros shipping rutorrent?

ruTorrent is not shipped very often with Linux Distributions. Arch Linux is anther distribution that ships it. There is a significant user base outside Linux Distributions, to the point where we can very effectively test and maintain the web client.

> i hate to admit that i was very close to proposing https://github.com/jesec/flood and ditching https://github.com/Novik/ruTorrent/ but for rutorrent we already have something going on in gentoo, and we can still add flood later, as alternative or replacement

We have a significantly higher user base than flood and require far fewer dependencies to run web client. I'm familiar with Gentoo Linux personally. ruTorrent would be a much better option for these reasons.

> Feel free to post a link to a PR or issue that we are waiting for, i assume you tell all your users and redistributors about that somewhere ...

The best way to contact us, is to file an issue report on the ruTorrent GitHub. As you can see currently only 8 out of 1972 issue reports are open currently. Very few torrent clients are able to achieve these level of efficiency out of the GitHub platform.

When I become a collaborator, we greatly improved our development processes and exponentially reduced the number of regressions being put into stable releases. Our new versioning system of stable releases is very suitable for Gentoo Linux!

**Comment by henning-schild** [comment]
@stickz that all sounds very good, will wait for you to get back and keep pushing this, and open an issue in the meantime on the git metadata in the release tarballs

File: www-apps/rutorrent/rutorrent-4.0.ebuild [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+	insinto "${MY_HTDOCSDIR}"
+	doins -r .
+
+	chmod +x "${ED}${MY_HTDOCSDIR}"/plugins/*/*.sh \
+		"$ED${MY_HTDOCSDIR}"/php/test.sh || die "chmod failed"
+
+	keepdir "${MY_HTDOCSDIR}"/share/settings
+	keepdir "${MY_HTDOCSDIR}"/share/torrents
+	keepdir "${MY_HTDOCSDIR}"/share/users
```
Comment (henning-schild) [inline www-apps/rutorrent/rutorrent-4.0.ebuild:50]: we also have "conf/users/" which we might have to keep

**Comment by stickz** [comment]
@henning-schild `ruTorrent` should be ready now. A hot fix is released to resolve the caching issues, when upgrading versions.
https://github.com/Novik/ruTorrent/releases/tag/v4.0.1-hotfix

All the JavaScript files are cache busted now when upgrading to version 4. Also the `WebUISettings` are no longer cached. This should directly resolve the linked bug reported by @CyberGuerro.

We would be happy to merge improvements for Linux distributions, if a pull request is created on the upstream repository. We would relay on a third-party because it's very difficult to test changes for Gentoo Linux. The testing environment setup time is high.

Also, the `cloudflare` plugin should have a use flag when deploying to Gentoo Linux. It creates an unnecessary dependency. It should not be "used" by default. Most users do not want to use this feature, due to the complexity of setup requirements.

**Comment by henning-schild** [comment]
> @henning-schild `ruTorrent` should be ready now. A hot fix is released to resolve the caching issues, when upgrading versions. https://github.com/Novik/ruTorrent/releases/tag/v4.0.1-hotfix
> 
> All the JavaScript files are cache busted now when upgrading to version 4. Also the `WebUISettings` are no longer cached. This should directly resolve the linked bug reported by @CyberGuerro.
> 
> We would be happy to merge improvements for Linux distributions, if a pull request is created on the upstream repository. We would relay on a third-party because it's very difficult to test changes for Gentoo Linux. The testing environment setup time is high.
> 
> Also, the `cloudflare` plugin should have a use flag when deploying to Gentoo Linux. It creates an unnecessary dependency. It should not be "used" by default. Most users do not want to use this feature, due to the complexity of setup requirements.

Thanks. I will update to that.

The "-stable" vs "-hotfix" in the tarball file name is really not very nice to have! I bet the next one will be "-stable" again. Just use the version and be done with it?

Also the browsers title line still shows 4.0 ... yes i did rule out caching. And i do not report it upstream because your issue templates are a pain in the neck.

**Comment by stickz** [comment]
> The "-stable" vs "-hotfix" in the tarball file name is really not very nice to have! I bet the next one will be "-stable" again. Just use the version and be done with it?
> 
> Also the browsers title line still shows 4.0 ... yes i did rule out caching. And i do not report it upstream because your issue templates are a pain in the neck.

Thank you for the feedback @henning-schild . This is valuable for us to improve our development processes.

1. Yes, we can remove the prefixes at the end of the version, to make it easier to distribute on Linux Distributions.

2. Yes, we can update the version line in the web browser for all minor point releases. We just started creating minor point releases to increase the stability of the web client. New features are still frozen. We want to make version 4.0 as stable as possible before version 4.1. There will be anther hot fix of version 4.0.2, when more bugs are reported to upstream.

3. I create a new issue template for development professionals you can use: https://github.com/Novik/ruTorrent/commit/0f9417a5cf6c2a1af33a5173e32587381b7dcc61 . The other issue templates are restrictive because we have users who "know very little about computers" or "barely know what a torrent client is" creating issue reports. They require strict guidance to create effective reports, which aid our development process.

**Comment by henning-schild** [comment]
pinging two people who recently worked on www-apps

@ajakk
@ceamac

File: www-apps/rutorrent/rutorrent-4.0.1.ebuild [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+pkg_setup() {
+	webapp_pkg_setup
+}
+
+src_install() {
+	webapp_src_preinst
+
+	rm -r .github || die
+	find . \( -name .gitignore -o -name .gitmodules \) -type f -exec rm '{}' + || die
```
Comment (ceamac) [inline www-apps/rutorrent/rutorrent-4.0.1.ebuild:37]:
```suggestion
	find . \( -name .gitignore -o -name .gitmodules \) -type f -delete || die
```

File: www-apps/rutorrent/rutorrent-4.0.1.ebuild [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	if [[ ${PV} == 9999 ]]; then
+		rm -r .git .gitattributes || die
+	fi
+
+	insinto "${MY_HTDOCSDIR}"
+	doins -r .
+
+	chmod +x "${ED}${MY_HTDOCSDIR}"/plugins/*/*.sh \
+		"$ED${MY_HTDOCSDIR}"/php/test.sh || die "chmod failed"
```
Comment (ceamac) [inline www-apps/rutorrent/rutorrent-4.0.1.ebuild:46]: Please add a comment here that fperms does not work with globbing
(and also `${ED}` for the second `$ED`, to be consistent)

**Review by ceamac — COMMENTED** [review]
I don't have much experience with webapps.  I installed this and it works in Apache.

It's a good thing to upgrade this, since the old version is EAPI 6.

Thank you for your work!

**Comment by stickz** [comment]
@henning-schild version 4.0.2 is released. The file is named `v4.0.2.tar.gz` as requested. The version display is updated in the web browser. A few minor bugs are fixed with the web application.

**Comment by stickz** [comment]
@henning-schild 
There are a few more optional dependencies recommended for ruTorrent:

1. The screenshots feature can utilize [media-video/ffmpeg](https://packages.gentoo.org/packages/media-video/ffmpeg) to take screenshots from video files.
2. The unpack feature can utilize [app-arch/unzip](https://packages.gentoo.org/packages/app-arch/unzip) and [app-arch/unrar](https://packages.gentoo.org/packages/app-arch/unrar) to decompress `.zip` and `.rar` files.

**Comment by henning-schild** [comment]
@stickz let us deal with such special details once that ebuild enters the tree

**Comment by stickz** [comment]
@henning-schild We're up to `v4.0.4` currently with `v4.1` right around the corner.

**Comment by henning-schild** [comment]
> @henning-schild We're up to `v4.0.4` currently with `v4.1` right around the corner.

Thanks for that reminder, i will look into it.

gentoo has some helpers for maintainers that are based on https://repology.org and https://repology.org/projects/?search=rutorrent now mentions 4.1. But i am not (yet) a maintainer of that so it does not show in https://packages.gentoo.org/maintainer/henning@hennsch.de/outdated

but it does show in https://packages.gentoo.org/maintainer/web-apps@gentoo.org/outdated

i will try to keep tracking that and might some day add myself as maintainer, but in case you do have an account for bugs.gentoo.org that would be the place to go, where you could add me to cc

**Pull request closed without merging**